bots/nymphora/nymphora_main.lua

```diff
@@ -0,0 +1,613 @@
+	local unitSelf = core.unitSelf
+	local myPos = unitSelf:GetPosition()
+
+	local bActionTaken = false
+
+	if behaviorLib.lastRetreatUtil > object.nStunThreshold and skills.stun:CanActivate() then
+		if core.NumberElements(core.localUnits.EnemyHeroes) > 0 then
+			local unitTarget = nil
+			local closestDistance = 999999999
+			for _, unit in pairs(core.localUnits.EnemyHeroes) do
+				local Distance2DSq = Vector3.Distance2DSq(myPos, unit:GetPosition())
+				if Distance2DSq < closestDistance then
+					unitTarget = unit
+					closestDistance = Distance2DSq
+				end
+			end
+
+			bActionTaken = core.OrderAbilityPosition(botBrain, skills.stun, unitTarget:GetPosition())
+		end
+	end
+
+	return bActionTaken
+end
+
+----------
+-- Push --
+----------
+function behaviorLib.customPushExecute(botBrain)
+	local unitSelf = core.unitSelf
+
+	bActionTaken = false
+
+	if unitSelf:GetManaPercent() > 0.7 then
+		local creeps = core.localUnits["EnemyCreeps"]
+		core.InsertToTable(creeps, core.localUnits["AllyCreeps"])
+
+		local vecCenterOfCreeps, nCreeps = core.GetGroupCenter(creeps)
+		--local centerOfCreeps = core.AoETargeting(unitSelf, skills.heal:GetRange(), 300, true, nil, nil, nil)
+		if nCreeps > 4 and vecCenterOfCreeps ~= nil then
+			if skills.heal:CanActivate() then
+				bActionTaken = useHeal(botBrain, vecCenterOfCreeps)
+			end
+			if not bActionTaken then
+				if skills.stun:CanActivate() then
+					bActionTaken = core.OrderAbilityPosition(botBrain, skills.stun, vecCenterOfCreeps)
+				end
+			end
+		end
+	end
+
+	return bActionTaken
+end
+
+-- Walk to own heal
+function behaviorLib.getHealedUtility(botBrain)
+	local time = HoN.GetGameTime()
+	if time - 1100 < object.nHealLastCastTime then
+		local unitSelf = core.unitSelf
+		if unitSelf:GetHealthPercent() < 0.95 then
+			local distance = Vector3.Distance2D(unitSelf:GetPosition(), object.vecHealPos) - 300
+			if distance < unitSelf:GetMoveSpeed() * (time - object.nHealLastCastTime) / 1000 then
+				return 40
+			end
+		end
+	end
+	return 0
+end
+
+function behaviorLib.getHealedExecute(botBrain)
+	return core.OrderMoveToPos(botBrain, core.unitSelf, object.vecHealPos)
+end
+
+behaviorLib.getHealedBehavior = {}
+behaviorLib.getHealedBehavior["Utility"] = behaviorLib.getHealedUtility
+behaviorLib.getHealedBehavior["Execute"] = behaviorLib.getHealedExecute
+behaviorLib.getHealedBehavior["Name"] = "Nymphora get healed"
+tinsert(behaviorLib.tBehaviors, behaviorLib.getHealedBehavior)
+
+
+----------------
+-- Use recall --
+----------------
+behaviorLib.heroRecallTarget = nil
+function behaviorLib.recallAllyUtility(botBrain)
+	if skills.recall:CanActivate() then
+		if core.NumberElements(core.localUnits["EnemyHeroes"]) == 0 and behaviorLib.lastRetreatUtil < 30 then
+			local vecMyPosition = core.unitSelf:GetPosition()
+			for _, hero in pairs(core.teamBotBrain.tAllyHeroes) do
+				if 2000 * 2000 < Vector3.Distance2DSq(vecMyPosition, hero:GetPosition()) then
+					if hero:GetHealthPercent() < 0.35 then

```

Comment:
This is an interesting one. This would work nicely with bots, however, this logic would be infuriating as a player - the logic if it's a player should look at the player's TimeToLiveUtilityFn, to determine if they need saving